File: projects/basis/sdk-review/py-v2/basis/modules/staking.py
import logging
from web3 import Web3
from .factory import load_abi

logger = logging.getLogger(__name__)


class StakingModule:
    def __init__(self, client, staking_address: str):
        self.client = client
        self.staking_address = Web3.to_checksum_address(staking_address)
        self.staking_abi = load_abi('AStasisVault.json')
        self.erc20_abi = load_abi('IERC20.json')
        self.contract = self.client.web3.eth.contract(address=self.staking_address, abi=self.staking_abi)

    def _approve_if_needed(self, token_address: str, spender: str, amount: int):
        if not self.client.account:
            raise ValueError("Wallet account is required for approval.")

        checksum_token = Web3.to_checksum_address(token_address)
        checksum_spender = Web3.to_checksum_address(spender)
        token_contract = self.client.web3.eth.contract(address=checksum_token, abi=self.erc20_abi)

        allowance = token_contract.functions.allowance(
            self.client.account.address, checksum_spender
        ).call()

        if allowance < amount:
            func = token_contract.functions.approve(checksum_spender, amount)
            tx = func.build_transaction({
                'from': self.client.account.address,
                'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            })
            signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
            tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
            self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

    def _build_and_send_tx(self, function_call):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")

        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)

        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _sync_tx(self, tx_hash: str):
        """Sync tx to backend. Non-fatal on failure."""
        try:
            if not tx_hash.startswith("0x"):
                tx_hash = "0x" + tx_hash
            self.client.api.sync_transaction(tx_hash)
        except Exception as e:
            logger.warning("Sync warning: %s", e)

    def buy(self, amount: int):
        """Wraps STASIS (MAINTOKEN) into wSTASIS. Auto-approves."""
        self._approve_if_needed(self.client.main_token_address, self.staking_address, amount)
        func = self.contract.functions.buy(amount)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def sell(self, shares: int, claim_usdb: bool = False, min_usdb: int = 0):
        """Unwraps wSTASIS back to STASIS, optionally converting to USDB."""
        func = self.contract.functions.sell(shares, claim_usdb, min_usdb)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def lock(self, shares: int):
        """Locks wSTASIS as collateral for borrowing. Auto-approves."""
        self._approve_if_needed(self.staking_address, self.staking_address, shares)
        func = self.contract.functions.lock(shares)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def unlock(self, shares: int):
        """Unlocks wSTASIS collateral."""
        func = self.contract.functions.unlock(shares)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def borrow(self, stasis_amount_to_borrow: int, days: int):
        """Pledges STASIS as collateral and borrows USDB against it.
        stasis_amount_to_borrow is the STASIS to pledge — USDB received is collateral value minus fees."""
        func = self.contract.functions.borrow(stasis_amount_to_borrow, days)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def repay(self):
        """Repays the active staking loan. Auto-approves USDB."""
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required.")
        usdb_contract = self.client.web3.eth.contract(
            address=Web3.to_checksum_address(self.client.usdb_address), abi=self.erc20_abi
        )
        balance = usdb_contract.functions.balanceOf(self.client.account.address).call()
        if balance > 0:
            self._approve_if_needed(self.client.usdb_address, self.staking_address, balance)
        func = self.contract.functions.repay()
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def extend_loan(self, days_to_add: int, pay_in_usdb: bool, refinance: bool):
        """Extends the active staking loan. Auto-approves USDB when pay_in_usdb is True."""
        if pay_in_usdb and self.client.account:
            usdb_contract = self.client.web3.eth.contract(
                address=Web3.to_checksum_address(self.client.usdb_address), abi=self.erc20_abi
            )
            balance = usdb_contract.functions.balanceOf(self.client.account.address).call()
            if balance > 0:
                self._approve_if_needed(self.client.usdb_address, self.staking_address, balance)
        func = self.contract.functions.extendLoan(days_to_add, pay_in_usdb, refinance)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def get_user_stake_details(self, user: str):
        """Returns (liquidShares, lockedShares, totalShares, totalAssetValue)."""
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.getUserStakeDetails(checksum_user).call()

    def get_available_stasis(self, user: str) -> int:
        """Gets available STASIS (collateral value minus pledged)."""
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.getAvailableStasis(checksum_user).call()

    def convert_to_shares(self, assets: int) -> int:
        """Converts STASIS amount to wSTASIS shares."""
        return self.contract.functions.convertToShares(assets).call()

    def convert_to_assets(self, shares: int) -> int:
        """Converts wSTASIS shares to STASIS amount."""
        return self.contract.functions.convertToAssets(shares).call()

    def total_assets(self) -> int:
        """Returns total STASIS held by the vault (available + pledged)."""
        return self.contract.functions.totalAssets().call()

    def add_to_loan(self, additional_stasis_to_borrow: int):
        """Adds to the existing staking loan by borrowing more."""
        func = self.contract.functions.addToLoan(additional_stasis_to_borrow)
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

    def settle_liquidation(self):
        """Settles a liquidation on the staking position."""
        func = self.contract.functions.settleLiquidation()
        result = self._build_and_send_tx(func)
        self._sync_tx(result['hash'])
        return result

File: projects/basis/sdk-review/py-v2/src/modules/Staking.ts
import { BasisClient } from '../BasisClient';
import AStasisVaultArtifact from '../abis/AStasisVault.json';
import IERC20Artifact from '../abis/IERC20.json';
import { Address } from 'viem';

export class StakingModule {
  private client: BasisClient;
  private stakingAddress: Address;

  constructor(client: BasisClient, stakingAddress: Address) {
    this.client = client;
    this.stakingAddress = stakingAddress;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  private async approveIfNeeded(tokenAddress: Address, spender: Address, amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required for approval.");
    }
    const account = this.client.walletClient.account;
    const allowance = await this.client.publicClient.readContract({
      address: tokenAddress,
      abi: IERC20Artifact.abi,
      functionName: 'allowance',
      args: [account.address, spender],
    }) as bigint;

    if (allowance < amount) {
      const { request } = await this.client.publicClient.simulateContract({
        account,
        address: tokenAddress,
        abi: IERC20Artifact.abi,
        functionName: 'approve',
        args: [spender, amount],
      });
      const hash = await this.client.walletClient.writeContract(request);
      await this.client.publicClient.waitForTransactionReceipt({ hash });
    }
  }

  /**
   * Wraps STASIS (MAINTOKEN) into wSTASIS.
   * Approves the staking contract to spend MAINTOKEN if needed.
   */
  async buy(amount: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Approve MAINTOKEN for staking contract
    await this.approveIfNeeded(this.client.mainTokenAddress, this.stakingAddress, amount);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'buy',
      args: [amount],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Unwraps wSTASIS back to STASIS, optionally converting to USDB.
   */
  async sell(shares: bigint, claimUSDB: boolean = false, minUSDB: bigint = 0n) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'sell',
      args: [shares, claimUSDB, minUSDB],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Locks wSTASIS as collateral for borrowing.
   */
  async lock(shares: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Approve staking contract to transfer wSTASIS (the staking contract IS the wSTASIS token)
    await this.approveIfNeeded(this.stakingAddress, this.stakingAddress, shares);

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'lock',
      args: [shares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Unlocks wSTASIS collateral.
   */
  async unlock(shares: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'unlock',
      args: [shares],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Pledges STASIS as collateral and borrows USDB against it.
   * The stasisAmountToBorrow parameter is the STASIS amount to pledge — USDB received is collateral value minus fees.
   */
  async borrow(stasisAmountToBorrow: bigint, days: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'borrow',
      args: [stasisAmountToBorrow, days],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Repays the active staking loan. Auto-approves USDB to the staking contract.
   */
  async repay() {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // Approve USDB to the staking contract for repayment
    const usdbBalance = await this.client.publicClient.readContract({
      address: this.client.usdbAddress,
      abi: IERC20Artifact.abi,
      functionName: 'balanceOf',
      args: [this.client.walletClient.account.address],
    }) as bigint;
    if (usdbBalance > 0n) {
      await this.approveIfNeeded(this.client.usdbAddress, this.stakingAddress, usdbBalance);
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'repay',
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Extends the active staking loan.
   */
  async extendLoan(daysToAdd: bigint, payInUSDB: boolean, refinance: boolean) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    // If paying in USDB, approve a generous amount to cover the extension fee
    if (payInUSDB) {
      const usdbBalance = await this.client.publicClient.readContract({
        address: this.client.usdbAddress,
        abi: IERC20Artifact.abi,
        functionName: 'balanceOf',
        args: [this.client.walletClient.account.address],
      }) as bigint;
      if (usdbBalance > 0n) {
        await this.approveIfNeeded(this.client.usdbAddress, this.stakingAddress, usdbBalance);
      }
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'extendLoan',
      args: [daysToAdd, payInUSDB, refinance],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Gets staking details for a user.
   * Returns [liquidShares, lockedShares, totalShares, totalAssetValue].
   */
  async getUserStakeDetails(user: Address) {
    return this.client.publicClient.readContract({
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'getUserStakeDetails',
      args: [user],
    });
  }

  /**
   * Gets the available STASIS (collateral value minus pledged).
   */
  async getAvailableStasis(user: Address) {
    return this.client.publicClient.readContract({
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'getAvailableStasis',
      args: [user],
    }) as Promise<bigint>;
  }

  /**
   * Converts STASIS amount to wSTASIS shares.
   */
  async convertToShares(assets: bigint) {
    return this.client.publicClient.readContract({
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'convertToShares',
      args: [assets],
    }) as Promise<bigint>;
  }

  /**
   * Converts wSTASIS shares to STASIS amount.
   */
  async convertToAssets(shares: bigint) {
    return this.client.publicClient.readContract({
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'convertToAssets',
      args: [shares],
    }) as Promise<bigint>;
  }

  /**
   * Returns total STASIS held by the vault (available + pledged).
   */
  async totalAssets(): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'totalAssets',
    }) as Promise<bigint>;
  }

  /**
   * Borrows additional STASIS against locked wSTASIS collateral on an existing loan.
   */
  async addToLoan(additionalStasisToBorrow: bigint) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'addToLoan',
      args: [additionalStasisToBorrow],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Settles a liquidated staking loan position.
   */
  async settleLiquidation() {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: this.stakingAddress,
      abi: AStasisVaultArtifact.abi,
      functionName: 'settleLiquidation',
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    this._syncTx(hash);
    return { hash, receipt };
  }
}
